cip/1.accepted/CIP2021-08-10-Operator-precedence.adoc

```diff
@@ -0,0 +1,727 @@
+
+However, the query string is a sequence of characters and not a tree.
+The parser turns the query string into such an operator tree.
+In the query string, nesting can clearly be denoted by parentheses.
+For instance, the query string
+----
+((3+2)*4)
+----
+allows a parser to unambiguously construct the operator tree shown above.
+
+For user convenience, better readability, and familiarity with common conventions in mathematical notation, openCypher (and most other query and programming languages) allows omitting the parentheses.
+For instance,
+----
+3+2*4
+----
+is also a valid expression.
+
+Without additional rules, however, it is not possible to unambiguously construct a operator tree for such an expression.
+In the particular case, two operator trees are possible:
+
+- `((3+2)*4)`
+- `(3+(2*4))`
+
+To avoid this kind of ambiguity, openCypher (and most other query and programming languages) has _precedence rules_.
+
+Precedence rules (or _rules of operator precedence_) define the order in which the different operators, i.e. the different kinds of expressions, are evaluated.
+For instance, numeric multiplication is evaluated before numeric addition, such that expression
+----
+3+2*4
+----
+is effectively evaluated as
+----
+(3+(2*4))
+----
+and, hence, results in `11` (rather than `20`).
+
+This CIP states the precedence rules of openCypher.
+
+== Proposal
+
+The precedence in openCypher is defined by (1) _precedence levels_ and (2) _operator associativity_.
+Both are defined in the following two subsections.
+
+=== Precedence levels
+
+The following <<precedenceLevels,table>> lists the precedence levels in descending order.
+
+.[[precedenceLevels]]Precedence levels
+[cols="<.<1a,<.<4a,<.<4a", options="header"]
+|===
+|Level         |Group         |Operators
+
+|12
+|Atoms
+|
+
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Literal[Literals]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Variable[Variables]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Parameter[Parameters]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=CaseExpression[Case expressions]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=ListComprehension[List comprehensions]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=PatternComprehension[Pattern comprehensions]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Reduce[Reduce operator]
+* quantifiers
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[Pattern predicates]
```

Comment:
For me a naked pattern appearing as an expression is a pattern predicate. The oC grammar does not have an extra production for pattern predicates, because these have the same syntax as the RelationshipsPattern non-terminal. However, the RelationshipsPattern non-terminal is not only used for pattern predicates but also in pattern comprehensions. Hence, we can not rename the RelationshipsPattern non-terminal to pattern predicate. 

In other words, when appearing as an atom a RelationshipsPattern is a (or plays the role of a) pattern predicate. Again, adding an extra production to the grammar would clarify this picture a bit.

Reply:
I see. I was more confused on the term "pattern predicates" because of GPMs "node pattern predicates" and "relationship pattern predicates". 

Is a naked pattern appearing as an expression generally referred to as "pattern predicate" in oC? It's easy to mix up with the GPM predicates.

Reply:
Well, generally. Terminology evolves with understanding and with people. It is in far a more recent term that we, i.e. the CLG, took the decision that naked patterns in expressions should evaluate to a boolean (rather than a list of paths, as they used to). It was argued, that patterns in match are also understood as predicates. So pattern elsewhere in the language should also behave like predicates. In general, it something evaluates to boolean it is reasonable to call it a predicates.

Anyway, it may be good to encode this name in the oC grammar. I am happy to do that. It is a conclusion that is somewhat similar to the one we have to quantifiers. :-)

Reply:
Done. I have also updated `CIP2015-05-13-EXISTS` to grammar refactorings ([rendered view](https://github.com/hvub/openCypher/blob/scenarios-for-precedence-rules/cip/1.accepted/CIP2015-05-13-EXISTS.adoc)), which provides a clear introduction point of the term "pattern predicate" now. It used the term already before, but now it does so much more prominently.